### ISSUE TITLE
How to support ADI (Arm Debug Interface) v5 in riscv?

### ISSUE THREAD

**Issue by coolwinner** [riscv/riscv-openocd#1000]
I checked the code, and I found it seems only MEM_AP is supported in the file mem_ap.c , but I can't find any files related with JTAG_AP.

Thanks!

**Comment by TommyMurphyTM1234** [comment]
Any use? Seems to have been an attempt to add JTAG_AP support.

* https://review.openocd.org/c/openocd/+/4875

But it seems to be still open after several years.

What exactly is the relevance of your query to RISC-V support?
Are there RISC-V implementations that use ADI (Arm Debug Interface) v5?

Maybe this is something that you need to raise on the upstream mailing lists or issue tracker?

* https://openocd.org/pages/mailing-lists.html
* https://openocd.org/pages/bug-tracker.html

**Comment by coolwinner** [comment]
Yes, the RISC-V implementation that use ADI (Arm Debug Interface) v5 is as below:

![image](https://github.com/riscv/riscv-openocd/assets/29885455/463d2d78-75f2-49e5-99af-3ce3af4e459a)

But openocd seems not support Dap in riscv target, could you give some suggestions?

Thanks!

**Comment by TommyMurphyTM1234** [comment]
> Yes, the RISC-V implementation that use ADI (Arm Debug Interface) v5 is as below:
> 
> ![image](https://private-user-images.githubusercontent.com/29885455/298608377-463d2d78-75f2-49e5-99af-3ce3af4e459a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU5MzUzNTEsIm5iZiI6MTcwNTkzNTA1MSwicGF0aCI6Ii8yOTg4NTQ1NS8yOTg2MDgzNzctNDYzZDJkNzgtNzVmMi00OWU1LTk5YWYtM2NlM2FmNGU0NTlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIyVDE0NTA1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiOGRlMWEwYzVmNDAzZDc2ODVjM2RkMTZlZGM5NWE1Mzk1OGNkZjVmZjk0MTRhZWNmMWRkZmNmMzUyNjI4N2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZNaKR9cZdsOHbNo-RsaD9R780vM7zkFE8srNzgBe7QQ)

What is this RISC-V implementation?

**Comment by coolwinner** [comment]
> > Yes, the RISC-V implementation that use ADI (Arm Debug Interface) v5 is as below:
> > ![image](https://private-user-images.githubusercontent.com/29885455/298608377-463d2d78-75f2-49e5-99af-3ce3af4e459a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU5MzUzNTEsIm5iZiI6MTcwNTkzNTA1MSwicGF0aCI6Ii8yOTg4NTQ1NS8yOTg2MDgzNzctNDYzZDJkNzgtNzVmMi00OWU1LTk5YWYtM2NlM2FmNGU0NTlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIyVDE0NTA1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiOGRlMWEwYzVmNDAzZDc2ODVjM2RkMTZlZGM5NWE1Mzk1OGNkZjVmZjk0MTRhZWNmMWRkZmNmMzUyNjI4N2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZNaKR9cZdsOHbNo-RsaD9R780vM7zkFE8srNzgBe7QQ)
> 
> What is this RISC-V implementation?

Sorry, I am afraid you may misunderstand my mean. I mean our hardware design is as the image I shown in above comment, but we want to debug with openocd, it seems current openocd not support the design. So could you give me some suggestions about how to support the hardware design with openocd?

Thanks!

**Comment by TommyMurphyTM1234** [comment]
According to the OpenOCD documentation:

* https://openocd.org/doc/html/Architecture-and-Core-Commands.html#:~:text=16.11%20RISC%2DV,implement%20version%200.11.

>  OpenOCD supports JTAG debug of RV32 and RV64 cores 

There is no mention of any ADI support.

It might be worth asking this question in the upstream OpenOCD mailing lists?

> How to support ADI (Arm Debug Interface) v5 in RISC-V

* https://openocd.org/pages/mailing-lists.html

Edit: perhaps work of this nature is relevant to your question?

* https://twitter.com/wren6991/status/1648754219875196944

**Comment by zqb-all** [comment]
hi @coolwinner  ,has there been any progress on this issue so far ? It seems that JTAG-AP is completely without any support, and even though openocd has some support for MEM-AP, it will require further modifications to support RISC-V DM behind MEM-AP

**Comment by coolwinner** [comment]
> hi @coolwinner ,has there been any progress on this issue so far ? It seems that JTAG-AP is completely without any support, and even though openocd has some support for MEM-AP, it will require further modifications to support RISC-V DM behind MEM-AP

Hi @zqb-all ,
We changed our design and no longer used JTAG-AP, and we are doing some modifications to support RISC-V DM behind MEM-AP.

**Comment by zqb-all** [comment]
> We changed our design and no longer used JTAG-AP, and we are doing some modifications to support RISC-V DM behind MEM-AP.

Got it. Thanks for your reply.

**Comment by ashig8** [comment]
> > hi @coolwinner ,has there been any progress on this issue so far ? It seems that JTAG-AP is completely without any support, and even though openocd has some support for MEM-AP, it will require further modifications to support RISC-V DM behind MEM-AP
> 
> Hi @zqb-all , We changed our design and no longer used JTAG-AP, and we are doing some modifications to support RISC-V DM behind MEM-AP.

Hello @coolwinner 
I am working on a similar design where RISCV-DM is behind MEM-AP. As you mentioned that you have done some modifications to support this, were you able to debug riscv ?  can you pls share some details on how this can be achieved ?
